### DISCUSSION TITLE
Do I still need the api-ms redirect DLLs in Windows 10/.NET 6?

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/runtime#65692] [Q&A] [answered]
Every application I publish as self-contained with .NET 6 and before outputs a large set of api-ms* assemblies/DLLs. While I understand for what reason they are here, I thought the original reason was that API call from former P/Invokes are redirected to newer API calls.

I assumed that increasing the RuntimeIdentifier to win10 or even higher versions, like net5.0-windows10.0.22000.0 gets rid of them but they are still there.

What is the reason why they are still there even in never version and can I safely delete them? I would also be nice if there would be a good reference to read about. I'm not sure if something is written in "Windows Internals" or if I missed it.

**Comment by jkotas** [comment]
You do not need the api-ms-redirect .dlls for Win10.

cc @VSadov

**Comment by jkotas** [comment, marked as answer]
> What is the reason why they are still there even in never version and can I safely delete them?

They are needed for Win7.

**Reply by VSadov** [reply]
"needed for Win7" is the current understanding, yes.

There is a good chance these files are no longer needed even on Win7 though, after we made winrt apiset to delay-load. (in 6.0, I think) 

We already do not include apiset files in singlefile apps, while the complete set of dependencies of a given app is the same regardless of the deployment mode. It is questionable why non-singlefile needs to be different.
Besides, after https://github.com/dotnet/runtime/pull/63196, singlefile apps work even on Windows7 (more precisely on Windows7 SP1 with UCRT installed).

I am looking at a way to remove these files cleanly, but initial attempt ran into some technical troubles. It is not too simple since the process how these files get published into a self-contained app is fairly convoluted and spans two repos (runtime and sdk). Once that figured out, we can experiment with various scenarios - to be sure, but I think we should be able to stop including these files.

Alternatively we can stop including them when targeting Win10+ RIDs. 
Although I believe we can drop them unconditionally just like in the singlefile case.

**Reply by VSadov** [reply]
I've logged https://github.com/dotnet/runtime/issues/65758 to track the removing of apisets (or realizing why they are still needed)

**Reply by msedi** [reply]
@VSadov: Thanks for the detailed explanation. That helps ;-)

**Comment by deeprobin** [comment]
Btw is there a way to disable these in csproj?

**Reply by jkotas** [reply]
You would have to write a custom target to delete them in a published app.

**Reply by deeprobin** [reply]
I think it would make sense if these files are not created at all (at least for build performance reasons - why copy into dir and then delete them?).

An option in the csproj file would not be bad in my opinion. The question is whether new features for R2R are still accepted, since there is now NativeAOT?

---

Edit: Oh thats not R2R spec. I think we really need a option to disable these files

**Reply by jkotas** [reply]
These files are gone in .NET 7. And this has nothing to do with R2R.
